src/blockdata/script.rs

```diff
@@ -632,113 +632,78 @@ pub enum Instruction<'a> {
 
 /// Iterator over a script returning parsed opcodes
 pub struct Instructions<'a> {
-    data: &'a [u8],
+    data: ::core::slice::Iter<'a, u8>,
     enforce_minimal: bool,
 }
 
 impl<'a> Iterator for Instructions<'a> {
     type Item = Result<Instruction<'a>, Error>;
 
     fn next(&mut self) -> Option<Result<Instruction<'a>, Error>> {
-        if self.data.is_empty() {
-            return None;
+        // we use macro because we need to `return`
+        macro_rules! next_push_data_len {
+            ($this:expr, $len:expr, $max:expr) => {
+                {
+                    let n = match read_uint_iter(&mut $this.data, $len) {
+                        Ok(n) => n,
+                        // We do exhaustive matching to not forget to handle new variants if we extend
+                        // `UintError` type.
+                        // Overflow actually means early end of script (script is definitely shorter
+                        // than `usize::max_value()`)
+                        Err(UintError::EarlyEndOfScript) | Err(UintError::NumericOverflow) => {
+                            $this.data.nth($this.data.len());  // Kill iterator so that it does not return an infinite stream of errors
+                            return Some(Err(Error::EarlyEndOfScript));
+                        },
+                    };
+                    if $this.enforce_minimal && n < $max {
+                        $this.data.nth($this.data.len());  // Kill iterator so that it does not return an infinite stream of errors
+                        return Some(Err(Error::NonMinimalPush));
+                    }
+                    let ret = Some(Ok(Instruction::PushBytes(&$this.data.as_slice()[..n])));
+                    $this.data.nth(n.max(1) - 1);
+                    ret
+                }
+            }
```

Comment:
Can this be a private function instead of a macro?

Reply:
Read the comment :)

Reply:
Actually, it may be doable.

Reply:
Done, thanks for hint! :)

Reply:
:)